### PR TITLE
feat!: deactivate localization component during the pose initialization

### PR DESCRIPTION
**Pull request by kminoda**
## Description
I will make this PR ready for review after taking an agreement inside TIER IV.

### Background
See also: https://github.com/orgs/autowarefoundation/discussions/2869

Current localization architecture has an issue when re-initializing the pose. If we want to re-initialize from point A to B, we cannot guarantee that both NDT and EKF states are updated to point B synchronously. This may end up in the localization estimation being pulled back to point A.
Screenshot from 2022-09-16 13-59-15

### Proposal
Here I make an assumption that the pose initialization is only performed when either of the following is met:
- the vehicle is stopped
- the vehicle is disengaged (or driven manually)

In other words, I do not consider the pose re-initialization when the vehicle is driving autonomously. (I believe that re-initialization during autonomous driving is rather a long-term challenge, so I leave it out of the scope for now)

To overcome this issue, I propose to deactivate the localization components (in current Universe, `ndt_scan_matcher` and `ekf_localizer`) during the pose initialization. When they are deactivated, the localization is expected to stop publishing the estimation. After the calculation finishes, the components throw away their pose&twist buffers they have and are activated. 

## Related links
- https://github.com/tier4/tier4_autoware_msgs/pull/54

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by yukkysaito** [comment]
discussion result with @kminoda 
memo
- This PR can be merged. But it is only a temporary measure.
- Permanent measures will be considered separately.
  - It is not right to have an error in initial position estimation since it is within the normal case.
  - Need to think about whether it is a good idea what pose_initializer have the global state management.

**Review by yukkysaito — APPROVED** [review]
LGTM

File: localization/pose_initializer/src/pose_initializer/localization_trigger_module.cpp [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  client_ekf_trigger_ = node->create_client<RequestTriggerNode>("ekf_trigger_node");
+  client_ndt_trigger_ = node->create_client<RequestTriggerNode>("ndt_trigger_node");
+}
+
+void LocalizationTriggerModule::deactivate() const
+{
+  const auto req = std::make_shared<RequestTriggerNode::Request>();
+  req->activate = false;
+  const auto res_ekf = client_ekf_trigger_->async_send_request(req).get();
+  const auto res_ndt = client_ndt_trigger_->async_send_request(req).get();
```
Comment (isamu-takagi) [inline localization/pose_initializer/src/pose_initializer/localization_trigger_module.cpp:30]: `res_ekf` and `res_ndt` are not used.

File: localization/pose_initializer/src/pose_initializer/localization_trigger_module.cpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+void LocalizationTriggerModule::activate() const
+{
+  const auto req = std::make_shared<RequestTriggerNode::Request>();
+  req->activate = true;
+  const auto res_ekf = client_ekf_trigger_->async_send_request(req).get();
+  const auto res_ndt = client_ndt_trigger_->async_send_request(req).get();
```
Comment (isamu-takagi) [inline localization/pose_initializer/src/pose_initializer/localization_trigger_module.cpp:38]: Should these be run sequentially? If not:
```suggestion
  const auto future_ekf = client_ekf_trigger_->async_send_request(req);
  const auto future_ndt = client_ndt_trigger_->async_send_request(req);
  future_ekf.wait();
  future_ndt.wait();
```

File: localization/pose_initializer/src/pose_initializer/pose_initializer_core.cpp [hunk 3]
```diff
@@ -44,6 +45,8 @@ PoseInitializer::PoseInitializer() : Node("pose_initializer")
     stop_check_duration_ = declare_parameter<double>("stop_check_duration");
     stop_check_ = std::make_unique<StopCheckModule>(this, stop_check_duration_ + 1.0);
   }
+  localization_trigger_ = std::make_unique<LocalizationTriggerModule>(this);
```
Comment (isamu-takagi) [inline localization/pose_initializer/src/pose_initializer/pose_initializer_core.cpp:48]: For planning simulation, this line should be enabled when `ndt_enabled = true`. And skip calling of activate/deactivate if this is nullptr.

Comment (kminoda) [reply localization/pose_initializer/src/pose_initializer/pose_initializer_core.cpp]: https://github.com/autowarefoundation/autoware.universe/pull/1873/commits/093494b80c373f32596cbbbc5aa7a9400cb0baa2
resolved

File: localization/pose_initializer/src/pose_initializer/localization_trigger_module.cpp (hunk 1 above)
Comment (kminoda) [reply localization/pose_initializer/src/pose_initializer/localization_trigger_module.cpp]: Thank you for all the comments!
https://github.com/autowarefoundation/autoware.universe/pull/1873/commits/093494b80c373f32596cbbbc5aa7a9400cb0baa2
resolved here

File: localization/pose_initializer/src/pose_initializer/localization_trigger_module.cpp (hunk 2 above)
Comment (kminoda) [reply localization/pose_initializer/src/pose_initializer/localization_trigger_module.cpp]: https://github.com/autowarefoundation/autoware.universe/commit/093494b80c373f32596cbbbc5aa7a9400cb0baa2
resolved here

**Comment by isamu-takagi** [comment]
@kminoda 
Build fails. Since get function is not a const, remove the const from the future type. And the wait function can be removed if the get function is used.
```
error: passing ‘const rclcpp::Client<std_srvs::srv::SetBool>::FutureAndRequestId’ as ‘this’ argument discards qualifiers [-fpermissive]
   48 |   if (future_ekf.get()->success & future_ndt.get()->success) {
```

**Comment by kminoda** [comment]
@isamu-takagi The build failure does not reproduce on my computer... (ubuntu 20.04, ROS 2 galactic).

**Comment by isamu-takagi** [comment]
@kminoda It occurs at humble. You can check the cause of the error in CI.
`build-and-test-differential / build-and-test-differential (humble, -cuda) (pull_request)`

**Comment by kminoda** [comment]
@isamu-takagi Thanks! I fixed the above issues.

**Review by isamu-takagi — APPROVED** [review]
LGTM

File: localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp [hunk 4]
```diff
@@ -860,3 +865,14 @@ std::optional<Eigen::Matrix4f> NDTScanMatcher::interpolateRegularizationPose(
   tf2::fromMsg(regularization_pose_msg.pose, regularization_pose_affine);
   return regularization_pose_affine.matrix().cast<float>();
 }
+
+void NDTScanMatcher::serviceTriggerNode(
+  const std_srvs::srv::SetBool::Request::SharedPtr req,
+  std_srvs::srv::SetBool::Response::SharedPtr res)
+{
+  if (req->data) {
+    initial_pose_msg_ptr_array_.clear();
```
Comment (YamatoAndo) [inline localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:874]:
```suggestion
    std::lock_guard<std::mutex> initial_pose_array_lock(initial_pose_array_mtx_);
    initial_pose_msg_ptr_array_.clear();
```

Comment (kminoda) [reply localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:875]: Thanks
resolved: https://github.com/autowarefoundation/autoware.universe/pull/1873/commits/e8713ac5be0f6647305bc59d78c6c4aca812815b

**Review by YamatoAndo — APPROVED** [review]

File: localization/ekf_localizer/src/ekf_localizer.cpp [hunk 5]
```diff
@@ -690,3 +693,21 @@ void EKFLocalizer::updateSimple1DFilters(const geometry_msgs::msg::PoseWithCovar
...
+  if (req->data) {
+    while (!current_pose_info_queue_.empty()) current_pose_info_queue_.pop();
+    while (!current_twist_info_queue_.empty()) current_twist_info_queue_.pop();
+    is_activated_ = true;
+  } else {
+    is_activated_ = false;
+  }
+  res->success = true;
+  return;
+}
```
Comment (xmfcx) [inline localization/ekf_localizer/src/ekf_localizer.cpp:713]: @kminoda this is a service, not a node ☹️

It is referenced in all the launch files as a node.

**Pull request merged**